crates/polars-plan/src/plans/optimizer/set_order/ir_pushdown.rs

```diff
@@ -0,0 +1,297 @@
+                let is_order_observing = !((function.has_equal_order() && all_outputs_unordered)
+                    || (function.is_input_order_agnostic()));
+                [is_order_observing].into()
+            },
+            IR::SimpleProjection { .. } => [!all_outputs_unordered].into(),
+            IR::Slice { .. } => [true].into(),
+            IR::HStack { input, exprs, .. } => {
+                let input = *input;
+                let mut observing = zip(exprs
+                    .iter()
+                    .map(|e| get_frame_observing(expr_arena.get(e.node()), expr_arena)));
+
+                let input_schema = ir_arena.get(input).schema(ir_arena).as_ref().clone();
+                ir = ir_arena.get_mut(node);
+                let IR::HStack { exprs, .. } = ir else {
+                    unreachable!()
+                };
+
+                let mut hits = 0;
+                for expr in exprs {
+                    hits += usize::from(input_schema.contains(expr.output_name()));
+                }
+
+                if hits < input_schema.len() {
+                    observing = adjust_for_with_columns_context(observing);
+                }
+
+                let is_order_observing =
+                    observing.is_none_or(|o| o.observes_frame() && !all_outputs_unordered);
+                [is_order_observing].into()
+            },
+            IR::Select { expr: exprs, .. } => {
+                let observing = zip(exprs
+                    .iter()
+                    .map(|e| get_frame_observing(expr_arena.get(e.node()), expr_arena)));
+                let is_order_observing =
+                    observing.is_none_or(|o| o.observes_frame() && !all_outputs_unordered);
+                [is_order_observing].into()
+            },
+
+            IR::Filter {
+                input: _,
+                predicate,
+            } => {
+                let observing = adjust_for_with_columns_context(get_frame_observing(
+                    expr_arena.get(predicate.node()),
+                    expr_arena,
+                ));
+                let is_order_observing =
+                    observing.is_none_or(|o| o.observes_frame() && !all_outputs_unordered);
+                [is_order_observing].into()
+            },
+
+            IR::Union { inputs, options } => {
+                if options.slice.is_none() && all_outputs_unordered {
+                    options.maintain_order = false;
+                }
+                std::iter::repeat_n(
+                    options.slice.is_some() || options.maintain_order,
+                    inputs.len(),
+                )
+                .collect()
+            },
+
+            IR::HConcat { inputs, .. } => std::iter::repeat_n(true, inputs.len()).collect(),
+
+            #[cfg(feature = "python")]
+            IR::PythonScan { .. } => UnitVec::new(),
+
+            IR::Sink { payload, .. } => [payload.maintain_order()].into(),
```

Comment:
Note to self. This needs to also keep the input if the partition expression observes order or if the partitioning strategy is `Parted`